python/pyarrow/_compute.pyx

```diff
@@ -2019,6 +2019,25 @@ cdef class Expression(_Weakrefable):
             return (<Expression> expr)
         return (<Expression> Expression._scalar(expr))
 
+    @staticmethod
+    def _call_function(str function_name, arguments,
+                       FunctionOptions options=None):
+        cdef:
+            vector[CExpression] c_arguments
+            shared_ptr[CFunctionOptions] c_options=(
+                <shared_ptr[CFunctionOptions]> nullptr)
+
+        for argument in arguments:
+            if not isinstance(argument, Expression):
+                raise TypeError("only other expressions allowed as arguments")
+            c_arguments.push_back((<Expression> argument).expr)
+
+        # if options is not None:
+        #     c_options = make_shared[CFunctionOptions](options.get_options())
```

Comment:
Given a python `FunctionOptions` class object, I need to get a shared pointer to the C++ object (`get_options()` returns a `const CFunctionOptions*`, but `CmakeCallExpressions` needs a `shared_ptr[CFunctionOptions]`). 

But I haven't directly found how to do that conversion (cc @pitrou)

Reply:
You should probably replace the `unique_ptr` in the FunctionOptions declaration with a `shared_ptr`.

Reply:
The  `DeserializeFunctionOptions` in C++ (in function_internal.h, and several other methods there) returns a `unique_ptr`(from https://github.com/apache/arrow/pull/10511). Can that be changed there as well?

Reply:
You can, but I don't think it's necessary on the C++ side. On the Python side, you can convert the unique_ptr into a shared_ptr (by moving it).

Reply:
Yes, that's what I did in the meantime. The only difficulty I ran into is that moving a uniqe_ptr to shared_ptr doesn't seem to work in cython, but with our helper `to_shared` it worked.